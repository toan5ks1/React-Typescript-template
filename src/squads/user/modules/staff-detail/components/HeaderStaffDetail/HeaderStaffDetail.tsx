import { Entities } from "src/common/constants/enum";
import { convertString } from "src/common/constants/helper";
import { Teacher, Staff } from "src/squads/user/service/bob/user-service-bob/types";

import { Box } from "@mui/material";
import Breadcrumbs from "src/components/Breadcrumbs";
import WrapperPageHeader from "src/components/Wrappers/WrapperPageHeader";

export interface HeaderStaffDetailProps {
    staff?: Teacher | Staff;
    isLoading?: boolean;
    onAction?: (...args: any[]) => void;
}

export const HeaderStaffDetail = ({ staff }: HeaderStaffDetailProps) => {
    let staffName: string | undefined;
    if (staff && "users" in staff) {
        staffName = staff?.users?.name;
    } else if (staff && "user" in staff) {
        staffName = staff?.user?.name;
    }

    return (
        <Box data-testid="HeaderStaffDetail">
            <Breadcrumbs resource={Entities.STAFF} name={staffName} />
            <WrapperPageHeader title={convertString(staffName)} />
        </Box>
    );
};
