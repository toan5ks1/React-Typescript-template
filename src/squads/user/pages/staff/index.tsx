import StaffList from "src/squads/user/modules/staff-list";

import WrapperPageContent from "src/components/Wrappers/WrapperPageContent";
import TranslationProvider from "src/squads/user/providers/TranslationProvider";

export default function StaffListPage() {
    return (
        <TranslationProvider>
            <WrapperPageContent>
                <StaffList />
            </WrapperPageContent>
        </TranslationProvider>
    );
}
