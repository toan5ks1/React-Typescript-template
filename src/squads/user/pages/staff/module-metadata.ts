import { Entities, Features } from "src/common/constants/enum";
import { IAppResource } from "src/models/resource";
import { MicroFrontendTypes } from "src/routing/type";

import { SchoolOutlined } from "@mui/icons-material";

const staff: IAppResource = {
    key: Entities.STAFF,
    name: Entities.STAFF,
    translateKey: "name",
    basename: `/${MicroFrontendTypes.USER}`,
    icon: SchoolOutlined,

    feature: Features.STAFF_MANAGEMENT,
};
export default staff;
