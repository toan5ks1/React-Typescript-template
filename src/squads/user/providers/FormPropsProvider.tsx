import { createContext, PropsWithChildren, useContext } from "react";

export interface FormPropsContextValue {
    readOnly: boolean;
}

const FormPropsContext = createContext<FormPropsContextValue>({ readOnly: false });

const { Provider } = FormPropsContext;

export const FormPropsProvider = ({
    children,
    readOnly,
}: PropsWithChildren<{ readOnly: boolean }>) => {
    return <Provider value={{ readOnly }}>{children}</Provider>;
};

export default FormPropsContext;
export const useFormProps = () => useContext(FormPropsContext);
