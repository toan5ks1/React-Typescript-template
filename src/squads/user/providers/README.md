### We abstract the provider into an isolated component, can be hard to test but parent component don't need to care about what the provider does
