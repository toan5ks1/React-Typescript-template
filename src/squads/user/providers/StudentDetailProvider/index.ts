export * from "./StudentDetailProvider";
export { default } from "./StudentDetailProvider";
