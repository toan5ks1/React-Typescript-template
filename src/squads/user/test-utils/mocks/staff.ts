import { ArrayElement } from "src/common/constants/types";
import {
    User_StaffListV2Query,
    User_StaffListV3Query,
    User_StaffOneV2Query,
    User_StaffListV4Query,
    User_UserGroupsManyReferenceV2Query,
} from "src/squads/user/service/bob/bob-types";
import { Staff, Teacher as TeacherOne } from "src/squads/user/service/bob/user-service-bob/types";
import { PaginationWithTotal } from "src/squads/user/service/service-creator";
import { userGroupsListToUserGroupIdsList } from "src/squads/user/service/usermgmt/staff-service-usermgmt";
import { NsUsermgmtStaffService } from "src/squads/user/service/usermgmt/staff-service-usermgmt/types";
import { CountryKeys } from "src/squads/user/typings/remote";

import { Country } from "manabuf/common/v1/enums_pb";
import { UserGroup } from "manabuf/usermgmt/v2/enums_pb";
import {
    CreateStaffResponse,
    UpdateStaffResponse,
    UpdateStaffSettingResponse,
} from "manabuf/usermgmt/v2/users_pb";

export const MOCK_STAFF_USERS_DATA = {
    name: "Thomas",
    email: "thomas@manabie.com",
    user_group: "USER_GROUP_TEACHER",
    avatar: "",
    country: "COUNTRY_JP",
    phone_number: "",
};
export interface Pagination {
    count: number;
    onPageChange: () => void;
    page: number;
    rowsPerPage: number;
    limit: number;
    offset: number;
    onRowsPerPageChange: () => void;
}

export interface TeacherInformation {
    teacher_by_school_ids: TeacherOne["teacher_by_school_ids"];
    users: TeacherOne["users"];
    teacher_id: TeacherOne["teacher_id"];
    id: TeacherOne["teacher_id"];
    country: CountryKeys;
}

export const mockPagination: Pagination = {
    count: 1,
    onPageChange: jest.fn(),
    page: 0,
    rowsPerPage: 10,
    limit: 10,
    offset: 0,
    onRowsPerPageChange: jest.fn(),
};

export const mockListStaffWithFilterV2: User_StaffListV2Query = {
    find_teacher_by_school_id: [...Array(5).keys()].map((value) => ({
        name: `Teacher ${value}`,
        email: `teach_email_${value}@gmail.com`,
        user_id: `user_id_${value}`,
        resource_path: "-2147483644",
    })),
    find_teacher_by_school_id_aggregate: {
        aggregate: {
            count: 5,
        },
    },
};

export const mockListStaffWithFilterV3: User_StaffListV3Query = {
    staff: [...Array(5).keys()].map((value) => ({
        staff_id: `staff_id_${value}`,
        user: {
            name: `Staff Name ${value}`,
            email: `staff_name_${value}@gmail.com`,
            user_group_members: [
                {
                    user_group: {
                        user_group_name: `Teacher ${value}`,
                        user_group_id: `user_group_id_${value}`,
                    },
                },
            ],
        },
    })),
    staff_aggregate: {
        aggregate: {
            count: 5,
        },
    },
};

export const mockListStaffWithFilterV4: User_StaffListV4Query = {
    staff: [...Array(5).keys()].map((value) => ({
        staff_id: `staff_id_${value}`,
        user: {
            name: `Staff Name ${value}`,
            email: `staff_name_${value}@gmail.com`,
            resource_path: "-2147483644",
            user_group_members: [
                {
                    user_group: {
                        user_group_name: `Teacher ${value}`,
                        user_group_id: `user_group_id_${value}`,
                    },
                },
            ],
        },
    })),
    staff_aggregate: {
        aggregate: {
            count: 5,
        },
    },
};

type StaffListQueryReturn = {
    data: User_StaffListV4Query["staff"];
    total: number;
};

export const mockListStaffReturnV4: StaffListQueryReturn = {
    data: [...Array(5).keys()].map((value) => ({
        staff_id: `staff_id_${value}`,
        user: {
            name: `Staff Name ${value}`,
            email: `staff_name_${value}@gmail.com`,
            resource_path: "-2147483644",
            user_group_members: [...Array(5).keys()].map((val) => ({
                user_group: {
                    user_group_name: `Teacher ${val}`,
                    user_group_id: `user_group_id_${val}`,
                },
            })),
        },
    })),
    total: 5,
};
interface StaffTableProp {
    data: User_StaffListV4Query["staff"];
    pagination: PaginationWithTotal;
    isFetching: boolean;
}

export const staffTableProp: StaffTableProp = {
    data: mockListStaffReturnV4.data,
    isFetching: false,
    pagination: mockPagination,
};

export const mockReturnStaffData = {
    data: mockListStaffWithFilterV4.staff,
    total: 1,
};

export const staffOneMockData: ArrayElement<User_StaffOneV2Query["staff"]> = {
    staff_id: "staff-id",
    user: {
        user_id: "Sierra",
        name: "Sierra",
        email: "sierra2022@gmail.com",
        user_group: "USER_GROUP_TEACHER",
        country: "COUNTRY_JP",
        user_group_members: [
            {
                user_group: {
                    user_group_id: "user_group_01",
                    user_group_name: "Student A",
                },
            },
            {
                user_group: {
                    user_group_id: "user_group_02",
                    user_group_name: "Student A",
                },
            },
        ],
    },
};

interface userStaffGetOneReturn {
    data: {
        staff: User_StaffOneV2Query["staff"];
    };
}

export interface StaffInformation {
    staff_id: Staff["staff_id"];
    user: ArrayElement<User_StaffOneV2Query["staff"]>["user"];
    country: CountryKeys;
}

export const mockUserStaffGetOneReturn: userStaffGetOneReturn = {
    data: {
        staff: [staffOneMockData],
    },
};

export const createMockStaff = (): StaffInformation => {
    return {
        staff_id: "staff-id",
        user: staffOneMockData.user,
        country: "COUNTRY_JP",
    };
};

export const userGroupsManyReference: User_UserGroupsManyReferenceV2Query["user_group"] = [
    ...Array(5).keys(),
].map((val) => ({
    user_group_name: `Teacher ${val}`,
    user_group_id: `user_group_id_${val}`,
}));

export const mockUseStaffInfoRulesReturn = {
    required: {
        value: true,
        message: "This field is required",
    },
    pattern: {
        email: {
            value: /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
            message: "Email address is not valid",
        },
    },
    validate: {
        email: jest.fn(),
    },
};

export const createStaffReq: NsUsermgmtStaffService.CreateStaffReq = {
    name: MOCK_STAFF_USERS_DATA.name,
    email: MOCK_STAFF_USERS_DATA.email,
    avatar: MOCK_STAFF_USERS_DATA.avatar,
    country: Country[MOCK_STAFF_USERS_DATA.country],
    phoneNumber: "",
    userGroup: UserGroup[MOCK_STAFF_USERS_DATA.user_group],
    organizationId: "1",
    userGroupsList: [...Array(5).keys()].map((val) => ({
        user_group_name: `Teacher ${val}`,
        user_group_id: `user_group_id_${val}`,
    })),
};

export const createStaffResp: NsUsermgmtStaffService.CreateStaffResp = {
    staff: {
        staffId: "staff_id",
        ...createStaffReq,
        userGroupIdsList: userGroupsListToUserGroupIdsList(createStaffReq.userGroupsList),
    },
};

export const updateStaffReq: NsUsermgmtStaffService.UpdateStaffReq = {
    staffId: "staff_id",
    name: MOCK_STAFF_USERS_DATA.name + " edited",
    email: MOCK_STAFF_USERS_DATA.email,
    userGroupsList: [],
};

export const updateStaffSettingReq = (
    autoCreateTimesheet: boolean
): NsUsermgmtStaffService.UpdateStaffSettingReq => {
    return {
        staffId: "staff_id",
        autoCreateTimesheet,
    };
};

export const createMockCreateStaffResponse = (): CreateStaffResponse => {
    const staffResp = new CreateStaffResponse.StaffProfile();
    staffResp.setStaffId("staff_id");
    staffResp.setName(MOCK_STAFF_USERS_DATA.name);
    staffResp.setEmail(MOCK_STAFF_USERS_DATA.email);
    staffResp.setAvatar(MOCK_STAFF_USERS_DATA.avatar);
    staffResp.setCountry(Country[MOCK_STAFF_USERS_DATA.country]);
    staffResp.setPhoneNumber("");
    staffResp.setUserGroup(UserGroup[MOCK_STAFF_USERS_DATA.user_group]);
    staffResp.setOrganizationId("1");

    const resp = new CreateStaffResponse();
    resp.setStaff(staffResp);
    return resp;
};

export const createMockUpdateStaffResponse = (): UpdateStaffResponse => {
    const resp = new UpdateStaffResponse();
    resp.setSuccessful(true);
    return resp;
};

export const createMockUpdateStaffSettingResponse = (): UpdateStaffSettingResponse => {
    const resp = new UpdateStaffSettingResponse();
    resp.setSuccessful(true);
    return resp;
};

export const createMockTeacherUsersData = (overrides: Partial<TeacherInformation["users"]>) => {
    return {
        ...MOCK_STAFF_USERS_DATA,
        ...overrides,
    };
};

export const createMockTeacher = (): TeacherInformation => {
    return {
        id: "teacher_id",
        teacher_id: "teacher_id",
        country: "COUNTRY_JP",
        users: MOCK_STAFF_USERS_DATA,
        teacher_by_school_ids: [
            {
                school_id: 1,
            },
        ],
    };
};
