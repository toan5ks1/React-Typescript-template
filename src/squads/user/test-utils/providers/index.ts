export { default as TestThemeProvider } from "./TestThemeProvider";
export { default as TestCommonAppProvider } from "./TestCommonAppProvider";
export * from "./TestQueryWrapper";
export { default as TestHookForm } from "./TestHookForm";
export { default as TestHookFormProvider } from "./TestHookFormProvider";
export { default as TestingRouter } from "./TestingRouter";
export { default as AppProvider } from "./AppProvider";
export { default as TestStudentDetailProvider } from "./TestStudentDetailProvider";
