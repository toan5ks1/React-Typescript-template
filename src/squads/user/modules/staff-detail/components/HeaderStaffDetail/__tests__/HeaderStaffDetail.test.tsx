import { staffOneMockData } from "src/squads/user/test-utils/mocks/staff";
import { TestCommonAppProvider } from "src/squads/user/test-utils/providers";

import { HeaderStaffDetail } from "../HeaderStaffDetail";

import { render, screen } from "@testing-library/react";

//TODO: Will update when we move component in phase - 2
jest.mock("src/hooks/useBreadcrumb", () => ({
    __esModule: true,
    default: () => ({
        breadcrumbs: [
            {
                url: `/staff`,
                name: "resources.staff.titles.staffManagement",
            },
        ],
    }),
}));

describe("<HeaderStaffDetail/>", () => {
    const renderComponent = () => {
        const wrapper = render(
            <TestCommonAppProvider>
                <HeaderStaffDetail staff={staffOneMockData} />
            </TestCommonAppProvider>
        );

        return wrapper;
    };
    it("should be to match snapshot", () => {
        const wrapper = renderComponent();

        expect(wrapper.container).toMatchSnapshot();
    });

    it("should render correct breadcrumb", () => {
        renderComponent();

        expect(screen.getByTestId("BreadcrumbItem")).toBeInTheDocument();
        expect(screen.getByTestId("BreadcrumbItem")).toHaveAttribute("href", "/staff");

        expect(screen.getByTestId("Breadcrumbs__entityName")).toBeInTheDocument();
        expect(screen.getByTestId("Breadcrumbs__entityName")).toHaveTextContent(
            staffOneMockData.user!.name
        );
    });
});
