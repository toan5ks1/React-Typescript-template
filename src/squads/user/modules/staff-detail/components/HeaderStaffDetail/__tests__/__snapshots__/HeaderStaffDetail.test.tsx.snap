// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderStaffDetail/> should be to match snapshot 1`] = `
.emotion-1 {
  margin-bottom: 4px;
}

.emotion-2 {
  margin: 0;
  font-size: 1rem;
  font-weight: 400;
  letter-spacing: 0.15px;
  text-transform: none;
  font-family: Roboto,sans-serif;
  line-height: 1.5;
  color: #757575;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0;
  margin: 0;
  list-style: none;
}

.emotion-4 {
  -webkit-text-decoration: unset;
  text-decoration: unset;
}

.emotion-5 {
  margin: 0;
  font-size: 0.75rem;
  font-weight: 400;
  letter-spacing: 0.4px;
  text-transform: none;
  font-family: Roboto,sans-serif;
  line-height: 1.66;
  color: #757575;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  margin-left: 8px;
  margin-right: 8px;
}

.emotion-7 {
  margin: 0;
  font-size: 0.75rem;
  font-weight: 400;
  letter-spacing: 0.4px;
  text-transform: none;
  font-family: Roboto,sans-serif;
  line-height: 1.66;
  color: #212121;
}

.emotion-8 {
  margin-bottom: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-10 {
  padding-bottom: 0px;
}

.emotion-11 {
  margin: 0;
  font-size: 1.25rem;
  font-weight: 500;
  letter-spacing: 0.15px;
  text-transform: none;
  font-family: Roboto,sans-serif;
  line-height: 1.6;
}

<div>
  <div
    class="MuiBox-root emotion-0"
    data-testid="HeaderStaffDetail"
  >
    <div
      class="MuiBox-root emotion-1"
    >
      <nav
        aria-label="breadcrumb"
        class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root emotion-2"
      >
        <ol
          class="MuiBreadcrumbs-ol emotion-3"
        >
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="emotion-4"
              data-testid="BreadcrumbItem"
              href="/staff"
            >
              <span
                class="MuiTypography-root MuiTypography-caption emotion-5"
                data-testid="TypographyShortenStr"
                title="resources.staff.titles.staffManagement"
              >
                resources.staff.titles.staffMa...
              </span>
            </a>
          </li>
          <li
            aria-hidden="true"
            class="MuiBreadcrumbs-separator emotion-6"
          >
            /
          </li>
          <li
            class="MuiBreadcrumbs-li"
          >
            <span
              class="MuiTypography-root MuiTypography-caption emotion-7"
              data-testid="Breadcrumbs__entityName"
            >
              Sierra
            </span>
          </li>
        </ol>
      </nav>
    </div>
    <div
      class="MuiBox-root emotion-8"
      data-testid="WrapperPageHeader__root"
    >
      <div
        class="MuiBox-root emotion-9"
      >
        <div
          class="MuiBox-root emotion-10"
        >
          <h6
            aria-label="title"
            class="MuiTypography-root MuiTypography-h6 emotion-11"
          >
            Sierra
          </h6>
        </div>
      </div>
      <div
        class="MuiBox-root emotion-0"
      />
    </div>
  </div>
</div>
`;
